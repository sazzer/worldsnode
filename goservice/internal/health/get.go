package health

import (
	"net/http"
	"strings"

	"github.com/unrolled/render"
)

var renderer *render.Render

func init() {
	renderer = render.New()
}

// HealthcheckDetailModel is the API Model representation of a single HealthCheck value
type healthcheckDetailModel struct {
	ObservedValue interface{} `json:"observedValue,omitempty"`
	ObservedUnit  string      `json:"observedUnit,omitempty"`
	Status        string      `json:"status"`
}

// HealthcheckModel is the API Model representation of the Healthcheck results
type healthcheckModel struct {
	Status  string                              `json:"status"`
	Details map[string][]healthcheckDetailModel `json:"details"`
}

// Get will perform the actual healthchecks and return the response
func (h *Healthchecker) get(w http.ResponseWriter, r *http.Request) {
	healthcheckResults := h.checkHealth()

	response := healthcheckModel{
		Status:  Pass,
		Details: make(map[string][]healthcheckDetailModel),
	}

	for _, healthcheckResult := range healthcheckResults {
		component := strings.Join([]string{healthcheckResult.Component, healthcheckResult.Measurement}, ":")

		response.Details[component] = append(response.Details[component], healthcheckDetailModel{
			Status:        healthcheckResult.Status,
			ObservedValue: healthcheckResult.Value,
			ObservedUnit:  healthcheckResult.Unit,
		})
		response.Status = combineHealth(response.Status, healthcheckResult.Status)
	}

	statusCode := 200
	if response.Status == Fail {
		statusCode = 500
	}

	renderer.JSON(w, statusCode, response)
}
