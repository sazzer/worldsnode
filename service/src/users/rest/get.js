// @flow

import {
    type $Request,
    type $Response,
} from 'express';
import {type UserRetriever} from '../service/retriever';
import {translateUserToModel} from './translator';
import pino from 'pino';

/** The logger to use */
const logger = pino();
/**
 * Get the details of a single user from the database
 *
 * @export
 * @param {$Request} req the request
 * @param {$Response} res the response
 * @param {UserRetriever} retriever The user retriever
 */
export default async function getUser(req: $Request, res: $Response, retriever: UserRetriever) {
    const userId = req.params.id;

    const user = await retriever.getById(userId);
    const userModel = translateUserToModel(user);
    logger.info({userModel}, 'Retrieved user');

    res
        .type('application/json')
        .status(200)
        .send(userModel);
}
