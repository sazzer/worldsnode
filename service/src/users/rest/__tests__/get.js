// @flow

import {Request, Response} from 'jest-express';
import * as testSubject from '../get';

describe('GET /api/users/123', () => {
    test('When the user exists', async () => {
        const user = {
            identity: {
                id: '123',
                version: '1',
                created: new Date('2018-11-15T09:34:00Z'),
                updated: new Date('2018-11-15T09:34:00Z'),
            },
            data: {
                name: 'Graham',
                logins: [],
            },
        };

        const request = new Request('/api/users/123');
        request.setParams({id: '123'});
        const response = new Response();
        const retriever = {
            getById: jest.fn(() => Promise.resolve(user)),
        };
        await testSubject.default(request, response, retriever);

        expect(response.type).toBeCalledWith('application/json');
        expect(response.status).toBeCalledWith(200);
        expect(response.send).toMatchSnapshot();
        expect(retriever.getById).toBeCalledWith('123');
    });
    test('When the user doesn\'t exist', () => {

    });
    test('When the ID is malformed', () => {

    });
    test('When the database is missing', () => {

    });
});
