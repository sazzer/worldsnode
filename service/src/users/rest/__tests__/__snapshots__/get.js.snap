// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/users/123 When the user exists 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "@context": "/jsonld/users/User.jsonld",
        "@id": "/api/users/123",
        "@type": Array [
          "Person",
          "User",
        ],
        "created": 2018-11-15T09:34:00.000Z,
        "email": undefined,
        "logins": Array [],
        "name": "Graham",
        "updated": 2018-11-15T09:34:00.000Z,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
